compiler_opt/tools/feature_importance.py

```diff
@@ -0,0 +1,271 @@
+# coding=utf-8
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""A tool for analyzing which features a model uses to make a decision.
+
+This script allows for processing a set of examples generated from a trace
+created through generate_default_trace into a set of shap values which
+represent how much that specific feature contributes to the final output of
+the model. These values can then be imported into an IPython notebook and
+graphed with the help of the feature_importance_graphs.py module in the same
+folder.
+
+Usage:
+PYTHONPATH=$PYTHONPATH:. python3 compiler_opt/tools/feature_importance.py \
+    --gin_files=compiler_opt/rl/regalloc/gin_configs/common.gin \
+    --gin_bindings=config_registry.get_configuration.implementation=\
+      @configs.RegallocEvictionConfig \
+    --data_path=/default_trace \
+    --model_path=/warmstart/saved_policy \
+    --num_examples=5 \
+    --output_file=./explanation_data.json
+
+The type of trace that is performed (ie if it is just tracing the default
+heuristic or if it is a trace of a ML model) doesn't matter as the only data
+that matters re the input features. The num_examples flag sets the number of
+examples that get processed into shap values. Increasing this value will
+potentially allow you to reach better conclusions depending upon how you're
+viewing the data, but increasing it will also increase the runtime of this
+script quite significantly as the process is not multithreaded.
+"""
+
+from absl import app
+from absl import flags
+from absl import logging
+import gin
+
+from compiler_opt.rl import data_reader
+from compiler_opt.rl import constant
+from compiler_opt.rl import registry
+
+import tensorflow as tf
+import shap
+import numpy
+import numpy.typing
+import json
+
+from tf_agents.typing import types
+from typing import Dict, Tuple
+
+SignatureType = Dict[str, Tuple[numpy.typing.ArrayLike, tf.dtypes.DType]]
+
+# only create flags if running as a script to prevent interference with pytest
+if __name__ == '__main__':
```

Comment:
We could avoid it if we refactored out all the data formatting functions into a separate module and then tested that rather than the script.

Reply:
Refactored all of the utilities into a separate file, `feature_importance_utils.py`, so that they can be unit tested independently of the script and we don't run into any flag conflicts.